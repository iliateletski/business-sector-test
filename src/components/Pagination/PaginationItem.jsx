import React from "react"
import styles from "./Pagination.module.css"
import { NavLink } from "react-router-dom"

const PaginationItem = ({children, to}) => {

    return (
        <li className={styles.pagination_item}>
            {
                typeof(children) === 'number'
                ? <NavLink
                className={({isActive}) => isActive ? styles.active : ''}
                to={`/${to}`}
            >
                {children}
                </NavLink>
                : <span>{children}</span>
            }
            
        </li>
    )
}

export default PaginationItem