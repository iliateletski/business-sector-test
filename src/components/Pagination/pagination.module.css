.pagination {
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 0 16px;
    flex: 0 0 auto;
}

.btn_go {
    font-size: 24px;
    line-height: 33px;
    background-color: transparent;
    border: none;
    color: #474955;
}

.pagination_item {
    color: #474955;
    font-size: 18px;
    font-weight: 700;
    line-height: 25px;
    font-style: italic;
    border: none;
    background-color: transparent;
}

.pagination_item +.pagination_item {
    margin-left: 14px;
}

.pagination_item .active {
    color: #7EBC3C;
}