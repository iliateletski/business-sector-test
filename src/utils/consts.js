export const TABLE_ROUTE = '/page'
export const LIMIT = 10
