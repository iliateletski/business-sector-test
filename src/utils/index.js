export const getPaginationList = (totalCount, limit, curentPage) => {
	const pagesTotal = Math.ceil(totalCount / limit)
	const pages = [1]

	if (curentPage > pagesTotal) {
		pages.push(2, 3, 4, 5)
	} else {
		let startPage = 2
		if (pagesTotal > 5 && curentPage > 3) {
			pages.push('...')
			startPage = Math.min(curentPage - 1, pagesTotal - 3)
		}
		const endPage = startPage + 4
		for (startPage; startPage !== endPage; startPage++) {
			pages.push(startPage)
		}
	}
	return { pages, pagesTotal }
}

const check = (value, query) => {
	if (typeof value !== 'string') return false
	return value.toLowerCase().includes(query.toLowerCase())
}

export const filterAndSort = (posts, sortParam, query) => {
	const filtered = query
		? posts.filter(
				post =>
					check(`${post.id}`, query) ||
					check(`${post.title}`, query) ||
					check(`${post.body}`, query)
		  )
		: [...posts]

	if (!filtered.length) return filtered

	if (sortParam) {
		return typeof filtered[0][sortParam] === 'number'
			? [...filtered].sort((a, b) => b[sortParam] - a[sortParam])
			: [...filtered].sort((a, b) => a[sortParam].localeCompare(b[sortParam]))
	}

	return filtered
}
