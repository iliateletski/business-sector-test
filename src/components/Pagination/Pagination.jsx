import React from 'react'
import { useSelector } from 'react-redux'
import { Link, useParams } from 'react-router-dom'

import { TABLE_ROUTE } from 'utils/consts'
import { LIMIT } from 'utils/consts'
import { getPaginationList } from 'utils/index'

import styles from './Pagination.module.css'
import PaginationItem from './PaginationItem'

const Pagination = () => {
	const totalCount = useSelector(state => state.totalCount)
	const { page } = useParams()
	const { pages, pageCount } = getPaginationList(totalCount, LIMIT, +page)

	return (
		<div className={styles.pagination} data-testid='Pagination'>
			<Link
				className={styles.link_go}
				to={`${TABLE_ROUTE}/${page > 1 ? page - 1 : page}`}
			>
				Назад
			</Link>
			<ul className={styles.pagination_items}>
				{pages.map(currentPage => (
					<PaginationItem key={currentPage} to={currentPage}>
						{currentPage}
					</PaginationItem>
				))}
			</ul>
			<Link
				className={styles.link_go}
				to={`${TABLE_ROUTE}/${page < pageCount ? +page + 1 : page}`}
			>
				Далее
			</Link>
		</div>
	)
}

export default Pagination
