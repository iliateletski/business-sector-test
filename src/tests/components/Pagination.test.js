import '@testing-library/jest-dom'
import { screen } from '@testing-library/react'
import * as reduxHooks from 'react-redux'
import { renderWithRouter } from 'tests/helpers/helpers'

import Pagination from 'components/Pagination/Pagination'

import * as functions from 'utils/index'

const testRoutes = {
	routes: [{ path: '/page/:page', element: <Pagination /> }],
	initialEntries: ['/page/1']
}

jest.mock('react-redux')
jest.mock('utils/index')

const mockUseSelector = jest.spyOn(reduxHooks, 'useSelector')
const mockGetPaginationList = jest.spyOn(functions, 'getPaginationList')
describe('Pagination component', () => {
	test('Page 1 active', () => {
		mockUseSelector.mockReturnValue(100)
		mockGetPaginationList.mockReturnValue({
			pages: [1, 2, 3, 4, 5],
			pageCount: 10
		})

		renderWithRouter(testRoutes)

		const pagination = screen.getByTestId('Pagination')
		expect(pagination).toMatchSnapshot()
	})

	test('Page 4 active', () => {
		mockUseSelector.mockReturnValue(100)
		mockGetPaginationList.mockReturnValue({
			pages: [1, '...', 3, 4, 5, 6],
			pageCount: 10
		})

		renderWithRouter({ ...testRoutes, initialEntries: ['/page/4'] })

		const pagination = screen.getByTestId('Pagination')
		expect(pagination).toMatchSnapshot()
	})

	test('Page 10 active', () => {
		mockUseSelector.mockReturnValue(100)
		mockGetPaginationList.mockReturnValue({
			pages: [1, '...', 7, 8, 9, 10],
			pageCount: 10
		})

		renderWithRouter({ ...testRoutes, initialEntries: ['/page/10'] })

		const pagination = screen.getByTestId('Pagination')
		expect(pagination).toMatchSnapshot()
	})
})
