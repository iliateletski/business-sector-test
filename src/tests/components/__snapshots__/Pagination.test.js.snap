// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component Page 1 active 1`] = `
<div
  class="pagination"
  data-testid="Pagination"
>
  <a
    class="link_go"
    href="/page/1"
  >
    Назад
  </a>
  <ul
    class="pagination_items"
  >
    <li
      class="pagination_item"
    >
      <a
        aria-current="page"
        class="active"
        href="/page/1"
      >
        1
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/2"
      >
        2
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/3"
      >
        3
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/4"
      >
        4
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/5"
      >
        5
      </a>
    </li>
  </ul>
  <a
    class="link_go"
    href="/page/2"
  >
    Далее
  </a>
</div>
`;

exports[`Pagination component Page 4 active 1`] = `
<div
  class="pagination"
  data-testid="Pagination"
>
  <a
    class="link_go"
    href="/page/3"
  >
    Назад
  </a>
  <ul
    class="pagination_items"
  >
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/1"
      >
        1
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <span>
        ...
      </span>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/3"
      >
        3
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        aria-current="page"
        class="active"
        href="/page/4"
      >
        4
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/5"
      >
        5
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/6"
      >
        6
      </a>
    </li>
  </ul>
  <a
    class="link_go"
    href="/page/5"
  >
    Далее
  </a>
</div>
`;

exports[`Pagination component Page 10 active 1`] = `
<div
  class="pagination"
  data-testid="Pagination"
>
  <a
    class="link_go"
    href="/page/9"
  >
    Назад
  </a>
  <ul
    class="pagination_items"
  >
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/1"
      >
        1
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <span>
        ...
      </span>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/7"
      >
        7
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/8"
      >
        8
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        class=""
        href="/page/9"
      >
        9
      </a>
    </li>
    <li
      class="pagination_item"
    >
      <a
        aria-current="page"
        class="active"
        href="/page/10"
      >
        10
      </a>
    </li>
  </ul>
  <a
    class="link_go"
    href="/page/10"
  >
    Далее
  </a>
</div>
`;
